### PR TITLE
HSEARCH-3556 Configuration of the max window size in Elasticsearch

### PR DESCRIPTION
**Pull request by fax4ever**
https://hibernate.atlassian.net/browse/HSEARCH-3556

### CODE REVIEW

File: backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/lowlevel/index/settings/impl/IndexSettings.java [hunk 1]
```diff
@@ -17,23 +17,30 @@
...
  */
 @JsonAdapter(IndexSettingsJsonAdapterFactory.class)
 public class IndexSettings {
 
+	public static final String MAX_RESULT_WINDOW_ATTRIBUTE = "max_result_window";
+
 	private Analysis analysis;
 
+	@SerializedName(IndexSettings.MAX_RESULT_WINDOW_ATTRIBUTE)
+	private Integer maxResultWindow;
```
Comment (yrodiere) [inline backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/lowlevel/index/settings/impl/IndexSettings.java:33]: I think you forgot to update the `merge` method?

File: backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/validation/impl/IndexSettingsValidator.java [hunk 2]
```diff
@@ -28,13 +28,14 @@
...
-			// No expectation
-			return;
+		if ( expectedAnalysis != null ) {
+			Analysis actualAnalysis = actual.getAnalysis();
+			validateAnalysisSettings( errorCollector, expectedAnalysis, actualAnalysis );
 		}
-		Analysis actualAnalysis = actual.getAnalysis();
 
-		validateAnalysisSettings( errorCollector, expectedAnalysis, actualAnalysis );
+		LeafValidators.EQUAL.validate( errorCollector, ValidationContextType.OTHER_INDEX_SETTINGS_ATTRIBUTE,
```
Comment (yrodiere) [inline backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/validation/impl/IndexSettingsValidator.java:36]: This should use `validateWithDefault`, I think? The settings returned by Elasticsearch will not necessarily include the value of `max_result_window_size`, in which case we should assume `10,000`.

One way to test this would be to create an index manually without specifying the `max_result_window_size`, then executing validation against that index. I'm fairly sure validation would fail with this code, but it should pass.

Probably a test to add to `ElasticsearchIndexSchemaManagerValidationOtherSettingsIT`?

File: documentation/src/main/asciidoc/reference/backend-elasticsearch.asciidoc [hunk 3]
```diff
@@ -891,6 +891,27 @@ refer to the documentation:
...
+----
+The default for this property is 10000.
+
+[NOTE]
+====
+This value will be used to limit the returning hits size if no limit has been defined on the query by the user.
+In this case if Hibernate Search notices there are more results, a warning will be issued.
+====
+
 == Custom index settings
```
Comment (yrodiere) [inline documentation/src/main/asciidoc/reference/backend-elasticsearch.asciidoc:915]: What happens if someone sets the max result window size through custom index settings? Do we take that into account? Is there a test for that?

File: documentation/src/main/asciidoc/reference/backend-elasticsearch.asciidoc [hunk 4]
```diff
@@ -891,6 +891,27 @@ refer to the documentation:
...
+hibernate.search.backend.searching.max_result_window_size = 5000
+# To configure a specific index:
+hibernate.search.backend.indexes.<index name>.searching.max_result_window_size = 5000
+----
+The default for this property is 10000.
+
+[NOTE]
+====
+This value will be used to limit the returning hits size if no limit has been defined on the query by the user.
+In this case if Hibernate Search notices there are more results, a warning will be issued.
```
Comment (yrodiere) [inline documentation/src/main/asciidoc/reference/backend-elasticsearch.asciidoc:912]:
```suggestion
In this case if Hibernate Search notices there are more results, a warning will be logged.
```

File: integrationtest/backend/elasticsearch/src/test/java/org/hibernate/search/integrationtest/backend/elasticsearch/mapping/ElasticsearchTypeNameMappingSchemaIT.java [hunk 5]
```diff
@@ -126,8 +128,17 @@ private JsonObject indexCreationPayload() {
...
 	}
+
+	private JsonObject getSettings() {
+		JsonObject settings = new JsonObject();
+		settings.add(
+				IndexSettings.MAX_RESULT_WINDOW_ATTRIBUTE,
+				new JsonPrimitive( ElasticsearchIndexSettings.Defaults.MAX_RESULT_WINDOW_SIZE )
+		);
+		return settings;
+	}
```
Comment (yrodiere) [inline integrationtest/backend/elasticsearch/src/test/java/org/hibernate/search/integrationtest/backend/elasticsearch/mapping/ElasticsearchTypeNameMappingSchemaIT.java:143]: ... Why is this necessary? You shouldn't have to set the defaults, they are implicit.

File: util/internal/integrationtest/backend/elasticsearch/src/main/java/org/hibernate/search/util/impl/integrationtest/backend/elasticsearch/rule/TestElasticsearchClient.java [hunk 6]
```diff
@@ -107,12 +109,23 @@ public boolean exists() {
...
+		public IndexClient deleteAndCreate(boolean withMaxResultWindow) {
+			TestElasticsearchClient.this.deleteAndCreateIndex( primaryIndexName, writeAlias, readAlias,
+					withMaxResultWindow );
 			return this;
 		}
 
 		public IndexClient deleteAndCreate(String settingsPath, String settings) {
-			JsonObject settingsAsJsonObject = buildStructuredSettings( settingsPath, settings );
+			// by default we add the withMaxResultWindow,
+			// so that we don't have any validation errors for this attribute
```
Comment (yrodiere) [inline util/internal/integrationtest/backend/elasticsearch/src/main/java/org/hibernate/search/util/impl/integrationtest/backend/elasticsearch/rule/TestElasticsearchClient.java:123]: Again, we shouldn't have validation errors when the max result window is the default. The validation code should assume that `null` means `10,000`. I left a comment on the validation code with a solution.

File: util/internal/integrationtest/backend/elasticsearch/src/main/java/org/hibernate/search/util/impl/integrationtest/backend/elasticsearch/rule/TestElasticsearchClient.java [hunk 7]
```diff
@@ -486,13 +489,26 @@ private JsonObject buildStructuredSettings(String settingsPath, String settings)
...
+				}
+
 				settingsJsonElement = parent;
 			}
 		}
 
 		return settingsJsonElement.getAsJsonObject();
 	}
 
+	private void addDefaultMaxResultWindow(JsonObject settingsJsonElement) {
```
Comment (yrodiere) [inline util/internal/integrationtest/backend/elasticsearch/src/main/java/org/hibernate/search/util/impl/integrationtest/backend/elasticsearch/rule/TestElasticsearchClient.java:505]: Again, we shouldn't have validation errors when the max result window is the default. The validation code should assume that `null` means `10,000`. I left a comment on the validation code with a solution.

File: documentation/src/main/asciidoc/reference/backend-elasticsearch.asciidoc (hunk 3 above)
Comment (yrodiere) [reply documentation/src/main/asciidoc/reference/backend-elasticsearch.asciidoc:915]: Actually... Shouldn't we rely exclusively on custom index settings, and remove the dedicated property? I think it would make sense, don't you?

File: backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/logging/impl/Log.java [hunk 8]
```diff
@@ -733,4 +733,9 @@ SearchException customIndexSettingsJsonSyntaxErrors(String filePath, @Cause Exce
 
 	@Message(id = ID_OFFSET + 139, value = "Unknown filter definition '%1$s'.")
 	SearchException unknownNamedPredicateForSearch(String name, @Param EventContext context);
+
+	@LogMessage(level = Level.WARN)
+	@Message(id = ID_OFFSET + 140, value = "Apply the default limit '%1$s' to the result hits size of '%2$s'." )
```
Comment (yrodiere) [inline backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/logging/impl/Log.java:738]:
```suggestion
	@Message(id = ID_OFFSET + 140, value = "A search query fetching all hits was requested, but only '%2$s' hits were retrieved because the maximum result window size forces a limit of '%1$s' hits. Refer to Elasticsearch's 'max_result_window_size' setting for more information." )
```

**Review by yrodiere — COMMENTED** [review]
A few comments... my main question is: do we really want a dedicated configuration property, when we could just retrieve the value of `max_result_window_size` from the custom settings?

**Comment by fax4ever** [comment]
I drastically reduced the scope of this contribution without introducing a new property for the setting.
@yrodiere feel free to review it again... thanks!

**Comment by fax4ever** [comment]
Here is a new full-build:https://ci.hibernate.org/blue/organizations/jenkins/hibernate-search/detail/PR-2540/4/.

File: backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/lowlevel/index/settings/impl/IndexSettings.java [hunk 9]
```diff
@@ -49,6 +52,15 @@ public void setAnalysis(Analysis analysis) {
 		return extraAttributes;
 	}
 
+	public Integer getMaxResultWindow() {
+		JsonElement maxResultWindowElement = extraAttributes.get( MAX_RESULT_WINDOW_ATTRIBUTE );
```
Comment (yrodiere) [inline backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/lowlevel/index/settings/impl/IndexSettings.java:56]: That's weird. Any reason to remove the more explicit `private Integer maxResultWindow`? The point of `extraAttributes` was to hold, well... extra attributes. The ones we don't care about.

File: backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/validation/impl/IndexSettingsValidator.java [hunk 10]
```diff
@@ -16,7 +19,7 @@
 
 import com.google.gson.JsonElement;
 
-public class IndexSettingsValidator implements Validator<IndexSettings> {
+	public class IndexSettingsValidator implements Validator<IndexSettings> {
```
Comment (yrodiere) [inline backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/validation/impl/IndexSettingsValidator.java:22]: Unnecessary tab here.

File: backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/lowlevel/index/settings/impl/IndexSettings.java (hunk 9 above)
Comment (yrodiere) [reply backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/lowlevel/index/settings/impl/IndexSettings.java:56]: Also, it seems to me this complicates `IndexSettingsValidator` unnecessarily.

File: util/internal/integrationtest/common/src/main/resources/log4j2.properties [hunk 11]
```diff
@@ -14,7 +14,7 @@ logger.jboss.level=info
 
 ### Search logger
 logger.search.name=org.hibernate.search
-logger.search.level=debug
+logger.search.level=trace
```
Comment (yrodiere) [inline util/internal/integrationtest/common/src/main/resources/log4j2.properties:17]: ?

File: backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/lowlevel/index/settings/impl/IndexSettings.java (hunk 9 above)
Comment (fax4ever) [reply backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/lowlevel/index/settings/impl/IndexSettings.java:56]: Ok I'm putting back the explicit field....

**Review by yrodiere — APPROVED** [review]
LGTM, feel free to merge once CI passes. Thanks!

**Review by yrodiere — APPROVED** [review]

**Comment by fax4ever** [comment]
Merged, thanks!
